src/Middleware/HttpLogging/src/FileLoggerProcessor.cs

```diff
@@ -0,0 +1,262 @@
+        {
+            // Files are grouped by day, and written up to _maxFileSize before rolling to a new file
+            DateTime today = DateTime.Now;
+            var fullName = GetFullName(today);
+            var fileInfo = new FileInfo(fullName);
+            if (!TryCreateDirectory())
+            {
+                // return early if we fail to create the directory
+                return;
+            }
+            var streamWriter = GetStreamWriter(fullName);
+
+            try
+            {
+                foreach (var message in messages)
+                {
+                    if (cancellationTokenSource.IsCancellationRequested)
+                    {
+                        return;
+                    }
+                    fileInfo.Refresh();
+                    // Roll to new file if _maxFileSize is reached
+                    // _maxFileSize could be less than the length of the file header - in that case we still write the first log message before rolling
+                    if (fileInfo.Exists && fileInfo.Length > _maxFileSize)
+                    {
+                        _fileNumber++;
+                        fullName = GetFullName(today);
+                        fileInfo = new FileInfo(fullName);
+                        streamWriter.Dispose();
+                        if (!TryCreateDirectory())
+                        {
+                            // return early if we fail to create the directory
+                            return;
+                        }
+                        streamWriter = GetStreamWriter(fullName);
+                    }
+                    if (!fileInfo.Exists || fileInfo.Length == 0)
+                    {
+                        await OnFirstWrite(streamWriter, _cancellationTokenSource);
+                    }
+
+                    await WriteMessageAsync(message, streamWriter, _cancellationTokenSource);
+                }
+            }
+            finally
+            {
+                RollFiles();
+                streamWriter.Dispose();
+            }
+
+        }
+
+        internal bool TryCreateDirectory()
+        {
+            if (!Directory.Exists(_path))
+            {
+                try
+                {
+                    Directory.CreateDirectory(_path);
+                    return true;
+                }
+                catch (Exception ex)
+                {
+                    Log.CreateDirectoryFailed(_logger, _path, ex);
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        // Virtual for testing
+        internal virtual async Task WriteMessageAsync(string message, StreamWriter streamWriter, CancellationTokenSource cancellationTokenSource)
+        {
+            if (cancellationTokenSource.IsCancellationRequested)
+            {
+                return;
+            }
+            await streamWriter.WriteLineAsync(message);
```

Comment:
Should be able to use https://source.dot.net/#System.Private.CoreLib/StreamWriter.cs,b6b9ec9473d4c200 to pass the cts, looks like FlushAsync might not have a cts overload though.

Reply:
https://github.com/dotnet/aspnetcore/pull/33251/commits/83050fd92fcf06353a109f83f27e0ce1430bc5d5